File: chat-go/internal/store/store.go
// Package store provides persistent, concurrent-safe storage for users and
// chat messages.  Data is written as JSON files inside a configurable directory.
package store

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"chat/internal/protocol"
)

// User is a registered account.
type User struct {
	ID           string    `json:"id"`
	Username     string    `json:"username"`
	PasswordHash string    `json:"password_hash"`
	CreatedAt    time.Time `json:"created_at"`
}

// Store holds users and messages in memory and persists them to disk.
// A sync.RWMutex protects the in-memory state so multiple goroutines can read
// concurrently while writes are serialised.
type Store struct {
	mu       sync.RWMutex
	users    map[string]*User          // keyed by lower-case username
	byID     map[string]*User          // keyed by user ID
	messages []*protocol.StoredMessage // ordered by insertion time
	dataDir  string
}

// New creates (or reopens) a Store backed by files in dataDir.
func New(dataDir string) (*Store, error) {
	if err := os.MkdirAll(dataDir, 0o755); err != nil {
		return nil, fmt.Errorf("store: create data dir: %w", err)
	}
	s := &Store{
		users:   make(map[string]*User),
		byID:    make(map[string]*User),
		dataDir: dataDir,
	}
	if err := s.load(); err != nil {
		return nil, err
	}
	return s, nil
}

// RegisterUser creates a new user account.  Returns an error when the username
// is already taken.
func (s *Store) RegisterUser(username, password string) (*User, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	key := strings.ToLower(username)
	if _, exists := s.users[key]; exists {
		return nil, fmt.Errorf("username %q is already taken", username)
	}

	u := &User{
		ID:           generateID(),
		Username:     username,
		PasswordHash: hashPassword(password),
		CreatedAt:    time.Now().UTC(),
	}
	s.users[key] = u
	s.byID[u.ID] = u
	return u, s.saveUsersLocked()
}

// Authenticate verifies credentials and returns the matching User.
func (s *Store) Authenticate(username, password string) (*User, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	u, ok := s.users[strings.ToLower(username)]
	if !ok {
		return nil, fmt.Errorf("user %q not found", username)
	}
	if u.PasswordHash != hashPassword(password) {
		return nil, fmt.Errorf("incorrect password")
	}
	return u, nil
}

// SaveMessage appends msg to the in-memory list and persists it to disk.
func (s *Store) SaveMessage(msg *protocol.StoredMessage) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.messages = append(s.messages, msg)
	return s.saveMessagesLocked()
}

// GetHistory returns the last n messages.  When n <= 0 all messages are
// returned.
func (s *Store) GetHistory(n int) []*protocol.StoredMessage {
	s.mu.RLock()
	defer s.mu.RUnlock()

	total := len(s.messages)
	if n <= 0 || n >= total {
		out := make([]*protocol.StoredMessage, total)
		copy(out, s.messages)
		return out
	}
	out := make([]*protocol.StoredMessage, n)
	copy(out, s.messages[total-n:])
	return out
}

// Search returns messages matching all non-empty criteria (AND logic):
//   - query    – case-insensitive substring match against content
//   - username – case-insensitive exact match against the sender's username
//   - from     – message timestamp must be >= from (inclusive)
//   - to       – message timestamp must be <= to   (inclusive)
func (s *Store) Search(query, username string, from, to *time.Time) []*protocol.StoredMessage {
	s.mu.RLock()
	defer s.mu.RUnlock()

	q := strings.ToLower(query)
	u := strings.ToLower(username)

	var out []*protocol.StoredMessage
	for _, m := range s.messages {
		if q != "" && !strings.Contains(strings.ToLower(m.Content), q) {
			continue
		}
		if u != "" && !strings.EqualFold(m.Username, u) {
			continue
		}
		if from != nil && m.Timestamp.Before(*from) {
			continue
		}
		if to != nil && m.Timestamp.After(*to) {
			continue
		}
		out = append(out, m)
	}
	return out
}

// ---------------------------------------------------------------------------
// internal helpers
// ---------------------------------------------------------------------------

func (s *Store) load() error {
	usersPath := filepath.Join(s.dataDir, "users.json")
	if data, err := os.ReadFile(usersPath); err == nil {
		var users []*User
		if err := json.Unmarshal(data, &users); err != nil {
			return fmt.Errorf("store: parse users.json: %w", err)
		}
		for _, u := range users {
			s.users[strings.ToLower(u.Username)] = u
			s.byID[u.ID] = u
		}
	}

	msgsPath := filepath.Join(s.dataDir, "messages.json")
	if data, err := os.ReadFile(msgsPath); err == nil {
		if err := json.Unmarshal(data, &s.messages); err != nil {
			return fmt.Errorf("store: parse messages.json: %w", err)
		}
	}
	return nil
}

func (s *Store) saveUsersLocked() error {
	users := make([]*User, 0, len(s.users))
	for _, u := range s.users {
		users = append(users, u)
	}
	return writeJSON(filepath.Join(s.dataDir, "users.json"), users)
}

func (s *Store) saveMessagesLocked() error {
	return writeJSON(filepath.Join(s.dataDir, "messages.json"), s.messages)
}

func writeJSON(path string, v any) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

func hashPassword(pw string) string {
	h := sha256.Sum256([]byte(pw))
	return hex.EncodeToString(h[:])
}

func generateID() string {
	// nano-timestamp + random hex nibbles — sufficient for a local demo.
	return fmt.Sprintf("%d-%04x", time.Now().UnixNano(), rand.Intn(0xFFFF))
}

File: chat-rust/src/store.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::RwLock;
use std::fs;

use anyhow::Result;
use chrono::{DateTime, Utc};
use hex;
use rand::Rng;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::protocol::StoredMessage;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub id: String,
    pub username: String,
    pub password_hash: String,
    pub created_at: DateTime<Utc>,
}

struct Inner {
    users: HashMap<String, User>,  // keyed by lowercase username
    by_id: HashMap<String, User>,  // keyed by user ID
    messages: Vec<StoredMessage>,
}

pub struct Store {
    inner: RwLock<Inner>,
    data_dir: PathBuf,
}

impl Store {
    pub fn new(data_dir: impl AsRef<Path>) -> Result<Self> {
        let data_dir = data_dir.as_ref().to_path_buf();
        fs::create_dir_all(&data_dir)?;

        let mut inner = Inner {
            users: HashMap::new(),
            by_id: HashMap::new(),
            messages: Vec::new(),
        };

        let users_path = data_dir.join("users.json");
        if users_path.exists() {
            let data = fs::read_to_string(&users_path)?;
            let users: Vec<User> = serde_json::from_str(&data)?;
            for u in users {
                inner.users.insert(u.username.to_lowercase(), u.clone());
                inner.by_id.insert(u.id.clone(), u);
            }
        }

        let msgs_path = data_dir.join("messages.json");
        if msgs_path.exists() {
            let data = fs::read_to_string(&msgs_path)?;
            inner.messages = serde_json::from_str(&data)?;
        }

        Ok(Self { inner: RwLock::new(inner), data_dir })
    }

    pub fn register_user(&self, username: &str, password: &str) -> Result<User> {
        let mut inner = self.inner.write().unwrap();
        let key = username.to_lowercase();

        if inner.users.contains_key(&key) {
            anyhow::bail!("username {:?} is already taken", username);
        }

        let user = User {
            id: generate_id(),
            username: username.to_string(),
            password_hash: hash_password(password),
            created_at: Utc::now(),
        };

        inner.users.insert(key, user.clone());
        inner.by_id.insert(user.id.clone(), user.clone());

        let users: Vec<User> = inner.users.values().cloned().collect();
        let path = self.data_dir.join("users.json");
        drop(inner);
        write_json(&path, &users)?;

        Ok(user)
    }

    pub fn authenticate(&self, username: &str, password: &str) -> Result<User> {
        let inner = self.inner.read().unwrap();
        let key = username.to_lowercase();

        let user = inner
            .users
            .get(&key)
            .ok_or_else(|| anyhow::anyhow!("user {:?} not found", username))?;

        if user.password_hash != hash_password(password) {
            anyhow::bail!("incorrect password");
        }

        Ok(user.clone())
    }

    pub fn save_message(&self, msg: StoredMessage) -> Result<()> {
        let mut inner = self.inner.write().unwrap();
        inner.messages.push(msg);
        let msgs = inner.messages.clone();
        let path = self.data_dir.join("messages.json");
        drop(inner);
        write_json(&path, &msgs)?;
        Ok(())
    }

    pub fn get_history(&self, n: usize) -> Vec<StoredMessage> {
        let inner = self.inner.read().unwrap();
        let total = inner.messages.len();
        if n == 0 || n >= total {
            inner.messages.clone()
        } else {
            inner.messages[total - n..].to_vec()
        }
    }

    pub fn search(
        &self,
        query: &str,
        username: &str,
        from: Option<DateTime<Utc>>,
        to: Option<DateTime<Utc>>,
    ) -> Vec<StoredMessage> {
        let inner = self.inner.read().unwrap();
        let q = query.to_lowercase();
        let u = username.to_lowercase();

        inner
            .messages
            .iter()
            .filter(|m| {
                if !q.is_empty() && !m.content.to_lowercase().contains(&q) {
                    return false;
                }
                if !u.is_empty() && m.username.to_lowercase() != u {
                    return false;
                }
                if let Some(from) = from {
                    if m.timestamp < from {
                        return false;
                    }
                }
                if let Some(to) = to {
                    if m.timestamp > to {
                        return false;
                    }
                }
                true
            })
            .cloned()
            .collect()
    }
}

fn hash_password(pw: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(pw.as_bytes());
    hex::encode(hasher.finalize())
}

fn generate_id() -> String {
    let ts = Utc::now().timestamp_nanos_opt().unwrap_or(0);
    let rand_part: u16 = rand::thread_rng().gen_range(0..0xFFFF);
    format!("{}-{:04x}", ts, rand_part)
}

fn write_json(path: &Path, v: &impl Serialize) -> Result<()> {
    let data = serde_json::to_string_pretty(v)?;
    fs::write(path, data)?;
    Ok(())
}
